core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -66,6 +68,8 @@ class LogSegment(val log: FileRecords,
   /* The maximum timestamp we see so far */
   @volatile private var maxTimestampSoFar = timeIndex.lastEntry.timestamp
   @volatile private var offsetOfMaxTimestamp = timeIndex.lastEntry.offset
+  /* The offset that would be used for the next message to be appended to this segment */
+  @volatile private var _nextOffset = INVALID_NEXT_OFFSET
```

Comment:
We can set initialize this to `baseOffset` in recovery because we will read all entries in `recovery()`. In the scenario that we do not call `recover()`, we need to initialize `nextOffset` to the `nextOffset` of the latest entry in the log segment file. To trigger this, we need to initialize `nextOffset` to -1 and call `nextOffsetFromLog()` the first time we access `nextOffset()`.